routing: Fix incorrect interface selection for egress pod routes

pchaigno opened the pull request:
The `Configure` method relies on the MAC address to select the proper egress interface for new pods (EKS and AKS). Several interfaces can however have the same MAC address, in which case the wrong interface may be selected.

To avoid this, we skip Linux slave devices during the lookup by MAC
address. Thus, in case of slave devices, we will select the master
device.

Fixes: https://github.com/cilium/cilium/pull/11073.

christarazi commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -67,7 +61,7 @@ func (info *RoutingInfo) Configure(ip net.IP, mtu int, compat bool) error {
 	if compat {
 		egressPriority = linux_defaults.RulePriorityEgress
-		tableID = ifindex
+		tableID = info.InterfaceNumber
```
This will break in Azure mode because the Azure API doesn't expose the ifindex. In Azure, the `InterfaceNumber` is hardcoded to to 0. https://docs.microsoft.com/en-us/rest/api/virtualnetwork/network-interfaces/get#networkinterface

christarazi reviewed (changes requested):
This will break Azure, but it can work for ENI mode.

pchaigno committed:
routing: Throw error if MAC lookup finds several devices

When setting up the Linux routes and rules for ENI and Azure, we lookup
the interfaces by their MAC addresses. In that case, we want to ensure a
single interface is found for the given MAC address. If several are
found, we throw an error now rather than to fail in a more obscure way
down the line.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
routing: Fix incorrect interface selection for pod routes

The Configure method relies on the MAC address to select the proper
egress interface for new pods (EKS and AKS). Several interfaces can
however have the same MAC address in the case of slave devices. In
such a case, the wrong interface may be selected.

To avoid this, we skip Linux slave devices during the lookup by MAC
address. Thus, in case of slave devices, we will select the master
device.

Fixes: 26308b63 ("Implement support for cilium-health in ENI mode")
Signed-off-by: Paul Chaignon <paul@cilium.io>

christarazi commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -278,35 +278,35 @@ func deleteRule(r route.Rule) error {
 	if err != nil {
-		err = fmt.Errorf("unable to list interfaces: %s", err)
-		return
+		return -1, fmt.Errorf("unable to list interfaces: %s", err)
```
While we're at it, let's make it `%w` instead of `%s`

christarazi commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -278,35 +278,35 @@ func deleteRule(r route.Rule) error {
-			return
-		}
+	if err = netlink.LinkSetMTU(link, mtu); err != nil {
+		return -1, fmt.Errorf("unable to change MTU of link %s to %d: %s", link.Attrs().Name, mtu, err)
```
Similarly here

christarazi commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -278,35 +278,35 @@ func deleteRule(r route.Rule) error {
+		return -1, fmt.Errorf("unable to change MTU of link %s to %d: %s", link.Attrs().Name, mtu, err)
+	}
+	if err = netlink.LinkSetUp(link); err != nil {
+		return -1, fmt.Errorf("unable to up link %s: %s", link.Attrs().Name, err)
```
And here

christarazi reviewed (approved):
Fix LGTM now, thanks! A few optional nits
